### PR TITLE
Virtual pipeline parallel support for LoRA in NLPAdapterModelMixin

### PR DESCRIPTION
**Pull request by vysarge**
# What does this PR do ?

Corrects handling of `model.virtual_pipeline_model_parallel_size` > 1 alongside `model.peft.peft_scheme=lora`; before this fix adapters are attached only to the first virtual pipeline chunk.

**Collection**: nlp

# Changelog 
- Replace parameter-based check for first pipeline stage with `parallel_state.is_pipeline_first_stage()` to account for virtual pipeline stage
- Alter various methods (`_get_all_keys`, `_check_and_add_peft_cfg`, `setup_optimizer_param_groups`, `set_tunable_base_params`, `tie_weights`, `get_peft_state_dict`, `load_state_dict`) to apply LoRA changes to all layers rather than only those in the first chunk and to write and read checkpoints correctly when virtual pipeline parallel is used

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py [hunk 1]
```diff
@@ -447,7 +463,11 @@ def sharded_state_dict(self, prefix: str = ''):
             return super().sharded_state_dict(prefix=prefix)
 
     def load_state_dict(self, state_dict, strict: bool = True):
-        if len(state_dict) == 0:
+        if len(state_dict) == 0 or (
+            parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None
+            and "model_0" in state_dict
+            and len(state_dict["model_0"]) == 0
```
[the remark this answers is not in the document]
Comment (vysarge) [reply nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py:472]: done!

**Review by erhoo82 — APPROVED** [review]
